go/arrow/compute/exprs/exec.go

```diff
@@ -0,0 +1,631 @@
+	if e.Root != expr.RootReference {
+		return nil, fmt.Errorf("%w: only RootReference is implemented", arrow.ErrNotImplemented)
+	}
+
+	ref, ok := e.Reference.(expr.ReferenceSegment)
+	if !ok {
+		return nil, fmt.Errorf("%w: only direct references are implemented", arrow.ErrNotImplemented)
+	}
+
+	expectedType, _, err := FromSubstraitType(e.GetType(), ext)
+	if err != nil {
+		return nil, err
+	}
+
+	var param compute.Datum
+	if sref, ok := ref.(*expr.StructFieldRef); ok {
+		if sref.Field < 0 || sref.Field >= int32(len(input.Values)) {
+			return nil, arrow.ErrInvalid
+		}
+		param = input.Values[sref.Field]
+		ref = ref.GetChild()
+	}
+
+	out, err := GetReferencedValue(compute.GetAllocator(ctx), ref, param, ext)
+	if err == compute.ErrEmpty {
+		out = compute.NewDatum(param)
+	} else if err != nil {
+		return nil, err
+	}
+	if !arrow.TypeEqual(out.(compute.ArrayLikeDatum).Type(), expectedType) {
+		return nil, fmt.Errorf("%w: referenced field %s was %s, but should have been %s",
+			arrow.ErrInvalid, ref, out.(compute.ArrayLikeDatum).Type(), expectedType)
+	}
+
+	return out, nil
+}
+
+func executeScalarBatch(ctx context.Context, input compute.ExecBatch, exp expr.Expression, ext ExtensionIDSet) (compute.Datum, error) {
+	if !exp.IsScalar() {
+		return nil, fmt.Errorf("%w: ExecuteScalarExpression cannot execute non-scalar expressions",
+			arrow.ErrInvalid)
+	}
+
+	switch e := exp.(type) {
+	case expr.Literal:
+		return literalToDatum(compute.GetAllocator(ctx), e, ext)
+	case *expr.FieldReference:
+		return execFieldRef(ctx, e, input, ext)
+	case *expr.Cast:
+		if e.Input == nil {
+			return nil, fmt.Errorf("%w: cast without argument to cast", arrow.ErrInvalid)
+		}
+
+		arg, err := executeScalarBatch(ctx, input, e.Input, ext)
+		if err != nil {
+			return nil, err
+		}
+		defer arg.Release()
+
+		dt, _, err := FromSubstraitType(e.Type, ext)
+		if err != nil {
+			return nil, fmt.Errorf("%w: could not determine type for cast", err)
+		}
+
+		var opts *compute.CastOptions
+		switch e.FailureBehavior {
+		case types.BehaviorThrowException:
+			opts = compute.SafeCastOptions(dt)
+		case types.BehaviorUnspecified:
+			opts = compute.UnsafeCastOptions(dt)
+		case types.BehaviorReturnNil:
+			return nil, fmt.Errorf("%w: cast behavior return nil", arrow.ErrNotImplemented)
+		}
+		return compute.CastDatum(ctx, arg, opts)
+	case *expr.ScalarFunction:
+		var (
+			err       error
+			allScalar = true
+			args      = make([]compute.Datum, e.NArgs())
+			argTypes  = make([]arrow.DataType, e.NArgs())
+		)
+		for i := 0; i < e.NArgs(); i++ {
+			switch v := e.Arg(i).(type) {
+			case types.Enum:
+				args[i] = compute.NewDatum(scalar.NewStringScalar(string(v)))
+			case expr.Expression:
+				args[i], err = executeScalarBatch(ctx, input, v, ext)
+				if err != nil {
+					return nil, err
+				}
+				defer args[i].Release()
+
+				if args[i].Kind() != compute.KindScalar {
+					allScalar = false
```

Comment:
Why do you care about `allScalar`?

Reply:
In the case where an expression is a scalar function whose arguments consist *solely* of enum values, we don't want the exec batch length for the input to that function to reflect the top level input. It's a special case where the input length for that evaluation is 1 since there's only the one value (the enum).

Reply:
The only case I can think of it being a valid function would be a generative function, i.e. a function that generates a result set itself like `SELECT RAND()` or whatnot, but that takes some enum arguments.